### ISSUE TITLE
Implement scrolling with mouse wheel

### ISSUE THREAD

**Issue by osa1** [osa1/tiny#130]

**Comment by trevarj** [comment]
What do we do now that we won't move to libtermkey or crossterm? Update term_input to read from /dev/input/mice? :laughing:

**Comment by osa1** [comment]
I think on most terminals (certainly the case on st and xterm) wheel-up generates the same byte sequence as `C-y` (just the byte `[25]`) and wheel-down generates the same byte sequence as `C-e`. We already capture `C-e` in `term_input` so we just need to map that to "scroll one line down" in the TUI. For `C-y` we need to map bytes `[25]` to `C-y` in term_input and similarly update TUI to scroll one line up on `C-y`. Easy!

**Comment by osa1** [comment]
Ahh, I remember the issue with this now. `C-e` is already mapped in TUI for moving the cursor to the end of the line (a key binding widely used by shells), so we can't use that (and wheel-down) for scrolling..

**Comment by trevarj** [comment]
In my term - gnome terminal - scrollup/scrolldown is mapped to ArrowUp and ArrowDown, so it just scrolls through the input history :confused:

**Comment by osa1** [comment]
Welp.. perhaps this needs some research then. Perhaps there's a way to enable mouse events in the terminals so that mouse events mapped to their own events rather than keyboard events. I briefly looked at crossterm, I think it doesn't generate mouse wheel events on Linux. Would be good to know how libtermkey handles mouse events.

**Comment by trevarj** [comment]
I think I may have a solution, if we want to copy how Termion does it.
They have these escape sequences that they write to `stdout` and it enables/disables terminal mouse control:
```
/// A sequence of escape codes to enable terminal mouse support.
const ENTER_MOUSE_SEQUENCE: &'static str = csi!("?1000h\x1b[?1002h\x1b[?1015h\x1b[?1006h");

/// A sequence of escape codes to disable terminal mouse support.
const EXIT_MOUSE_SEQUENCE: &'static str = csi!("?1006l\x1b[?1015l\x1b[?1002l\x1b[?1000l");
```
The csi! macro just prepends the CSI escape `\x1B[`.

On my terminal I entered:
`echo "\x1B[?1000h\x1b[?1002h\x1b[?1015h\x1b[?1006h" >> /dev/stdout`
and it disabled the scrolling, so when I do a `less -f /dev/stdin` I am no longer getting the up/down arrow byte sequence, but the correct one that we can parse in term_input :smiley: 

![image](https://user-images.githubusercontent.com/5448324/82823819-e87a8e00-9eb0-11ea-9da9-8c6473f72eae.png)

**Comment by osa1** [comment]
Yeah I think you found a solution. We had a similar sequence to enable focus gained/lost events, though I can't find that code now. It's possible that I deleted it a while ago (those events were never used).

term_input doesn't have access to the tty so I think we'll have to write those sequences in termbox.

**Comment by trevarj** [comment]
Sadly, those are xterm and rxvt escape codes to turn on/off mouse capture. Nothing specific to which buttons.

xterm ones can be found here: https://www.xfree86.org/4.8.0/ctlseqs.html#Mouse%20Tracking

```
 CSI ? P m h      DEC Private Mode Set (DECSET)
 
                  P s = 1 0 0 0 → Send Mouse X & Y on button press and release. See the section Mouse Tracking.
                  P s = 1 0 0 1 → Use Hilite Mouse Tracking.
                  P s = 1 0 0 2 → Use Cell Motion Mouse Tracking.
```

**Comment by osa1** [comment]
I think this is not easily possible to implement, and it's a minor convenience that is not worth adding extra maintenance burden. See #201 for what we tried. In summary:

- If we don't enable mouse events in the terminal then wheel-down is mapped to `C-e`, but we that key is already mapped to "move cursor to the end of the line".
- If we enable mouse events then terminal stops handling selection so we'll have to implement text selection for copying as well. That is too much effort for such a small feature that we already support with `shift-up`/`shift-down` and `PageUp`/`PageDown` keys.

Closing as wontfix.
